File: buzu/src/test/java/br/net/buzu/pplspec/exception/PplExceptionTest.java
package br.net.buzu.pplspec.exception;

import static org.junit.Assert.*;

import org.junit.Test;

/**
 * @author Douglas Siviotti
 *
 */
public class PplExceptionTest {

	private static void assertPplException(PplException e, String message, Class<? extends Throwable> causeClass) {
		if (message != null) {
			assertEquals(message, e.getMessage());
		}
		if (causeClass != null) {
			assertEquals(causeClass, e.getCause().getClass());
		}

	}

	private static void assertPplException(PplException e, Class<? extends Throwable> causeClass) {
		assertPplException(e, null, causeClass);
	}

	private static void assertPplException(PplException e, String message) {
		assertPplException(e, message, null);
	}

	@Test
	public void testPplException() {
		assertPplException(new PplException("msg"), "msg");
		assertPplException(new PplException(new RuntimeException("msgRuntime")), RuntimeException.class);
		assertPplException(new PplException("msg", new RuntimeException("msgRuntime")), "msg", RuntimeException.class);
	}

	@Test
	public void testPplParseException() {
		PplParseException e = new PplParseException("meta", "msg", String.class);
		assertPplException(new PplParseException("msg"), "msg");
		assertPplException(new PplParseException(e), PplParseException.class);
		assertPplException(new PplParseException("msg", new RuntimeException("msgRuntime")), "msg",
				RuntimeException.class);
	}

	@Test
	public void testPplSerializeException() {
		assertPplException(new PplSerializeException("msg"), "msg");
		assertPplException(new PplSerializeException(new RuntimeException("msgRuntime")), RuntimeException.class);
		assertPplException(new PplSerializeException("msg", new RuntimeException("msgRuntime")), "msg",
				RuntimeException.class);
	}

	@Test
	public void testPplReflectionException() {
		assertPplException(new PplReflectionException("method", String.class, new RuntimeException()), RuntimeException.class);
		assertPplException(new PplReflectionException("msg"), "msg");
		assertPplException(new PplReflectionException(new RuntimeException("msgRuntime")), RuntimeException.class);
		assertPplException(new PplReflectionException("msg", new RuntimeException("msgRuntime")), "msg",
				RuntimeException.class);
	}

	@Test
	public void testPplTypeMismatchException() {
		PplTypeMismatchException e = new PplTypeMismatchException(String.class, "string", Integer.class, Byte.class);
		assertTrue(e.getMessage().contains("string"));
		assertPplException(new PplTypeMismatchException("msg"), "msg");
		assertPplException(new PplTypeMismatchException(new RuntimeException("msgRuntime")), RuntimeException.class);
		assertPplException(new PplTypeMismatchException("msg", new RuntimeException("msgRuntime")), "msg",
				RuntimeException.class);
	}

	@Test
	public void testPplMetaclassViolationException() {
		assertPplException(new PplMetaclassViolationException("msg"), "msg");
	}

	@Test
	public void testPplConstraintViolationException() {
		PplConstraintViolationException e = new PplConstraintViolationException("red", "[white,black]");
		assertTrue(e.getMessage().contains("red"));
		assertTrue(e.getMessage().contains("[white,black]"));
		assertPplException(new PplConstraintViolationException("msg"), "msg");
	}

}

File: ppl-spec/src/test/kotlin/br/net/buzu/pplspec/exception/PplExceptionTest.kt
package br.net.buzu.pplspec.exception

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

/**
 * @author Douglas Siviotti
 */
class PplExceptionTest {

    private fun assertPplException(e: PplException, message: String?, causeClass: Class<out Throwable>? = null) {
        if (message != null) {
            assertEquals(message, e.message)
        }
        if (causeClass != null) {
            if (e.cause != null) {
                assertEquals(causeClass, e.cause?.javaClass)
            }
        }

    }

    private fun assertPplException(e: PplException, causeClass: Class<out Throwable>) {
        assertPplException(e, null, causeClass)
    }

    @Test
    fun testPplException() {
        assertPplException(PplException("msg"), "msg")
        assertPplException(PplException(RuntimeException("msgRuntime")), RuntimeException::class.java)
        assertPplException(PplException("msg", RuntimeException("msgRuntime")), "msg", RuntimeException::class.java)
    }

    @Test
    fun testPplParseException() {
        val e = PplParseException("meta", "msg", String::class.java)
        assertPplException(PplParseException("msg"), "msg")
        assertPplException(PplParseException(e), PplParseException::class.java)
        assertPplException(PplParseException("msg", RuntimeException("msgRuntime")), "msg",
                RuntimeException::class.java)
    }

    @Test
    fun testPplSerializeException() {
        assertPplException(PplSerializeException("msg"), "msg")
        assertPplException(PplSerializeException(RuntimeException("msgRuntime")), RuntimeException::class.java)
        assertPplException(PplSerializeException("msg", RuntimeException("msgRuntime")), "msg",
                RuntimeException::class.java)
    }

    @Test
    fun testPplReflectionException() {
        assertPplException(PplReflectionException("method", String::class.java, RuntimeException()), RuntimeException::class.java)
        assertPplException(PplReflectionException("msg"), "msg")
        assertPplException(PplReflectionException(RuntimeException("msgRuntime")), RuntimeException::class.java)
        assertPplException(PplReflectionException("msg", RuntimeException("msgRuntime")), "msg",
                RuntimeException::class.java)
    }

    @Test
    fun testPplTypeMismatchException() {
        val e = PplTypeMismatchException(String::class.java, "string", Int::class.java, Byte::class.java)
        val message = e.message ?: ""
        assertTrue(message.contains("string"))
        assertPplException(PplTypeMismatchException("msg"), "msg")
        assertPplException(PplTypeMismatchException(RuntimeException("msgRuntime")), RuntimeException::class.java)
        assertPplException(PplTypeMismatchException("msg", RuntimeException("msgRuntime")), "msg",
                RuntimeException::class.java)
    }

    @Test
    fun testPplMetaclassViolationException() {
        assertPplException(PplMetaclassViolationException("msg"), "msg")
    }

    @Test
    fun testPplConstraintViolationException() {
        val e = PplConstraintViolationException("red", "[white,black]")
        val message = e.message ?: ""
        assertTrue(message.contains("red"))
        assertTrue(message.contains("[white,black]"))
        assertPplException(PplConstraintViolationException("msg"), "msg")
    }

}
